### PR TITLE
feat: grype explain

### PR DESCRIPTION
**Pull request by willmurphyscode**
Fixes https://github.com/anchore/grype/issues/1342.

~~Still WIP. Mostly using https://github.com/anchore/grype/issues/1342 to comment on what the desired output should look like, then will clean up this branch and get ready to merge.~~

This is an initial prototype of `grype explain`. The goal is to get something in front of users so that they can try out the experience and we can start getting feedback. This is NOT a final version of the output, or a final factoring of the code that renders the output.

Note: this is invoked on grype JSON output, like ` grype -q alpine:3.7 -o json | go run cmd/grype/main.go explain --id CVE-2022-48174`.

### CODE REVIEW

File: grype/presenter/explain/__snapshots__/explain_snapshot_test.snap [hunk 1]
```diff
@@ -0,0 +1,108 @@
+
+[TestExplainSnapshot/keycloak-CVE-2020-12413 - 1]
+CVE-2020-12413 from nvd:cpe (Medium)
```
Comment (spiffcs) [inline grype/presenter/explain/__snapshots__/explain_snapshot_test.snap:3]: Output for the first iteration of this LGTM

File: grype/presenter/explain/explain.go [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+
+type ViewModel struct {
+	PrimaryVulnerability   models.VulnerabilityMetadata
+	RelatedVulnerabilities []models.VulnerabilityMetadata
+	MatchedPackages        []*explainedPackage // I think this needs a map of artifacts to explained evidence
+	URLs                   []string
+}
+
+type viewModelBuilder struct {
+	// TODO: this field is unused. Need think a bit more here.
```
Comment (spiffcs) [inline grype/presenter/explain/explain.go:34]: Do we want to remove this field if it's unused now? Are there still design aspects being thought through for exposing this in a different way?

File: grype/presenter/explain/explain.go [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+type VulnerabilityExplainer interface {
+	ExplainByID(IDs []string) error
+	ExplainBySeverity(severity string) error
+	ExplainAll() error
+}
+
+type ViewModel struct {
+	PrimaryVulnerability   models.VulnerabilityMetadata
+	RelatedVulnerabilities []models.VulnerabilityMetadata
+	MatchedPackages        []*explainedPackage // I think this needs a map of artifacts to explained evidence
```
Comment (spiffcs) [inline grype/presenter/explain/explain.go:29]: Is this comment still valid?

File: cmd/grype/cli/legacy/explain.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+			decoder := json.NewDecoder(os.Stdin)
+			err := decoder.Decode(&parseResult)
+			if err != nil {
+				return fmt.Errorf("unable to parse piped input: %+v", err)
+			}
+			explainer := explain.NewVulnerabilityExplainer(os.Stdout, &parseResult)
+			return explainer.ExplainByID(cveIDs)
+		}
+		// perform a scan, then explain requested CVEs
+		// TODO: implement
```
Comment (spiffcs) [inline cmd/grype/cli/legacy/explain.go:40]: I'm ok keeping this comment as a fast follow for non piped inputs

File: grype/presenter/explain/explain.go [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+	RelatedMatches       []models.Match
+	requestedIDs         []string // the vulnerability IDs the user requested explanations of
+}
+
+type Findings map[string]ViewModel
+
+type explainedPackage struct {
+	PURL                string
+	Name                string
+	Version             string // TODO: is there only going to be one of these?
```
Comment (spiffcs) [inline grype/presenter/explain/explain.go:46]: Is this comment  still valid? Did we find cases where there are other versions?

File: grype/presenter/explain/explain.go [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+	}
+	for _, m := range doc.Matches {
+		for _, related := range m.RelatedVulnerabilities {
+			key := related.ID
+			existing, ok := builders[key]
+			if !ok {
+				existing = newBuilder(requestedIDs)
+				builders[key] = existing
+			}
+			// TODO: need to pass in info about the related vulnerability
```
Comment (spiffcs) [inline grype/presenter/explain/explain.go:131]: Is this TODO still needed?

File: grype/presenter/explain/explain.go [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+		if b.PrimaryMatch.Vulnerability.ID != "" {
+			b.WithRelatedMatch(b.PrimaryMatch)
+		}
+		b.WithPrimaryMatch(m)
+	} else {
+		b.WithRelatedMatch(m)
+	}
+}
+
+// TODO: is this still needed?
```
Comment (spiffcs) [inline grype/presenter/explain/explain.go:162]: Based on reading this function I think it's still needed, but maybe you have found a way to simplify the `WithMatch` branches where we don't need to care about `Primary` vs `Related` anymore?

**Review by spiffcs — COMMENTED** [review]
Comments added about cleaning up the comments and potential reminders that have been sitting around - are there any integration/cli tests we want to add for this one. Maybe just a smoke test to give the 👍 it's executing and producing the correct results?

The rest of explain looks technically correct - no hard to catch bugs or otherwise hard to read code. 

@willmurphyscode I can do another pass this PM so we can get this prototype out for users to start kicking the tires on

**Review by wagoodman — APPROVED** [review]
This is a great v0 of this functionality, really looking forward to continuing this work!

**Review by spiffcs — APPROVED** [review]

**Pull request merged**